### PR TITLE
rewrite Buffer as a GraphStage

### PR DESCRIPTION
**Pull request by Hawstein**
Refs #21528

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
-      if (buffer.isEmpty) ctx.pushAndFinish(elem)
-      else ctx.push(elem)
-    } else if (ctx.isHoldingUpstream) ctx.pushAndPull(buffer.dequeue())
-    else if (buffer.isEmpty) ctx.holdDownstream()
-    else ctx.push(buffer.dequeue())
-  }
+    override def beforePreStart(): Unit = {
+      super.beforePreStart()
+      pull(in)
+    }
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: why not just in `preStart`?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+        case Backpressure ⇒ elem ⇒
+          buffer.enqueue(elem)
+          if (!buffer.isFull) pull(in)
+        case Fail ⇒ elem ⇒
+          if (buffer.isFull) failStage(new BufferOverflowException(s"Buffer overflow (max capacity was: $size)!"))
+          else {
+            buffer.enqueue(elem)
+            pull(in)
+          }
+      }
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Vals up top thanks!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+    }
+
+    override def onPull(): Unit = {
+      if (isClosed(in)) {
+        push(out, buffer.dequeue())
+        if (buffer.isEmpty) {
+          completeStage()
         }
+      } else if (!(isClosed(in) || hasBeenPulled(in))) {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: `!isClosed` is implied here because of the `if(isClosed` above?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 4]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+        push(out, buffer.dequeue())
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Do we know the buffer isn't empty here?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 5]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+      if (isAvailable(out)) {
+        push(out, elem)
+        pull(in)
+      } else {
+        enqueueAction(elem)
+      }
+    }
+
+    override def onPull(): Unit = {
+      if (isClosed(in)) {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I'd like the most common case first rather than this. One idea would to even have a separate `OutHandler` and use that after in has been closed.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 6]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+        push(out, buffer.dequeue())
+        if (buffer.isEmpty) {
+          completeStage()
         }
+      } else if (!(isClosed(in) || hasBeenPulled(in))) {
+        push(out, buffer.dequeue())
+        pull(in)
+      } else if (buffer.nonEmpty) {
+        push(out, buffer.dequeue())
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: We pushed an element from the buffer, so it cannot be full and in has not been pulled, shouldn't we pull?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 7]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
-final case class Buffer[T](size: Int, overflowStrategy: OverflowStrategy) extends DetachedStage[T, T] {
+final case class Buffer[T](size: Int, overflowStrategy: OverflowStrategy) extends SimpleLinearGraphStage[T] {
 
-  private var buffer: BufferImpl[T] = _
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
 
-  override def preStart(ctx: LifecycleContext): Unit = {
-    buffer = BufferImpl(size, ctx.materializer)
-  }
+    var buffer: BufferImpl[T] = _
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: buffer should be a val and created directly in constructor, is never replaced only cleared

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 8]
```diff
@@ -705,61 +705,81 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
-          buffer.enqueue(elem)
-          ctx.pull()
+    override def preStart(): Unit = {
+      buffer = BufferImpl(size, materializer)
+    }
+
+    override def onPush(): Unit = {
+      val elem = grab(in)
+      if (isAvailable(out)) {
+        push(out, elem)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: So, `out` will only be available when the buffer contains elements? If not, shouldn't it dequeue-push elements from the buffer if there are elements in it?

**Review by johanandren — COMMENTED** [review]
More of a set of questions than an actual review.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 7 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: It is in the constructor though? Of the GraphStageLogic.

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Yes, but the actual instance is created in `preStart`, and is mutable, so this would be better as a val and instantiated right here.

**Comment by patriknw** [comment]
Looks like there are some corner cases to be hardened here. Since tests are passing that is an indication that it would be good to add some more test coverage also.

**Comment by ktoso** [comment]
Hey @Hawstein, would you be able to address the recent comments here?

**Comment by Hawstein** [comment]
@ktoso Sorry for the long time delay, too busy to reply your comments and update the PR. Will fix it now :)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 8 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Good catch. I just simply translated the old version to GraphStage. Ashamed for not even thinking about it.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 6 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Nope. There already exists an pending pull which implied by the above `else if`

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 2 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Sure

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 7 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: @johanandren It does not work. It is not allowed defined the buffer as `val` in GraphStageLogic constructor. The compiler will complain.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Put it in preStart now

**Comment by Hawstein** [comment]
@ktoso @patriknw @johanandren The PR is updated.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 4 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: No, will fix it.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 7 above)
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Why would the compiler complain about that?
`val buffer = BufferImpl(size, materializer)`

In general val + mutable datastructure or var + immutable data structure is good to strive for if you have to do mutable stuff.

**Review by johanandren — APPROVED** [review]
LGTM with the buffer a val and instantiated in the constructor

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 9]
```diff
@@ -842,61 +842,80 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+    }
+
+    override def onPush(): Unit = {
+      val elem = grab(in)
+      if (isAvailable(out)) {
+        if (buffer.isEmpty) {
+          push(out, elem)
+          pull(in)
+        } else {
+          push(out, buffer.dequeue())
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: how is this correct? if `isAvailable(out)` would be true, then we were already pulled. But in that case we should have already delivered a dequeued element from the buffer and hence `isAvailable(out)` cannot be true.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 10]
```diff
@@ -842,61 +842,80 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+          enqueueAction(elem)
         }
+      } else {
+        enqueueAction(elem)
+      }
     }
 
+    override def onPull(): Unit = {
+      if (isClosed(in)) {
+        if (buffer.nonEmpty) push(out, buffer.dequeue())
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]:
```scala
if (buffer.isEmpty) push(out, ...)
else completeStage()
```

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 11]
```diff
@@ -842,61 +842,80 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+          push(out, buffer.dequeue())
+          enqueueAction(elem)
         }
+      } else {
+        enqueueAction(elem)
+      }
     }
 
+    override def onPull(): Unit = {
+      if (isClosed(in)) {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: why not factor out all `if (buffer.nonEmpty) push(out, buffer.dequeue())` calls to a single line as the first thing in the method, and then dispatch on `if(isClosed(in)) ... else if(!hasBeenPulled(in))`

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 12]
```diff
@@ -842,61 +842,80 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
...
+        if (buffer.nonEmpty) push(out, buffer.dequeue())
+        pull(in)
+      } else if (buffer.nonEmpty) {
+        push(out, buffer.dequeue())
+      }
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (buffer.isEmpty) completeStage()
+      else if (isAvailable(out)) onPull()
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: how can this happen? If the buffer is not empty, then `isAvailable(out)` cannot be true unless we were pulled and forgot to push a dequeued element.

**Review by drewhk — CHANGES_REQUESTED** [review]
Almost there, but there are some logic discrepancies remaining that need to be ironed out.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 9 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: `isAvailable(out)` can be true here when it has been pulled but there is no element in buffer to delivered. I have written some comments in the source file.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 11 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Refactor as you suggested. Thanks:)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 12 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: You are right. Have deleted it .

**Comment by Hawstein** [comment]
@drewhk  The PR is updated.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 9 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: > isAvailable(out) can be true here when it has been pulled but there is no element in buffer to delivered. I have written some comments in the source file.

Ok, but then there must have been a push() in-between, otherwise buffer.dequeue() would fail. But if there was a push, then that element should have been emitted, not buffered.

To put it shortly:
 - isAvailable(out) implies that there were a pull() before which was not answered
 - buffer.nonEmpty implies that there was a push() whose element has not been dequeued
 - but these two together are a contradiction

Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: This is not a problem since I have updated the code:

```scala
    override def onPush(): Unit = {
       val elem = grab(in)
       // If out is available, then it has been pulled but no dequeued element has been delivered.
       // It means the buffer at this moment is definitely empty,
       // so we just push the current element to out, then pull.
       if (isAvailable(out)) {
         push(out, elem)
         pull(in)
       } else {
         enqueueAction(elem)
       }
     }
```

**Review by drewhk — APPROVED** [review]
LGTM

**Comment by Hawstein** [comment]
Does the test will run automatically again? It failed because of the `pr-validation:test`, I think it does nothing to do with the unit tests.

**Comment by johanandren** [comment]
PLS BUILD

**Comment by Hawstein** [comment]
Can this PR be merged since there are 2 LGTM 😄

**Comment by johanandren** [comment]
Thanks @Hawstein !
